Update marrmot01 class

SarahAlidoost opened the pull request.

SarahAlidoost committed:
add available_versions

SarahAlidoost committed:
refactor init based on new API

SarahAlidoost committed:
refactor notebook based on new API

SarahAlidoost committed:
remove parameterset from the constructor, add default values for parameters

SarahAlidoost committed:
add versions and container images

SarahAlidoost committed:
re-run notebook locally

SarahAlidoost committed:
revise docstrings

SarahAlidoost committed:
refactor init

SarahAlidoost committed:
remove marrmot.singularity_image and marrmot.docker_image from config file

sverhoeven committed:
Got example Marrmot notebook to work + more tests + small cleanings

sverhoeven committed:
Make mypy happy

SarahAlidoost committed:
fix a bug: remove initial values from setup()

SarahAlidoost committed:
re-run the example notebook

SarahAlidoost committed:
fix links, revise docstrings, replace some double quotes with single one

SarahAlidoost committed:
fix test: remove initial values from setup in tests

Peter9192 commented on ewatercycle/models/marrmot.py:
```diff
@@ -42,42 +42,70 @@ def _generate_work_dir(work_dir: PathLike = None):
 
 class MarrmotM01(AbstractModel):
-    """eWaterCycle implementation of Marrmot Collie River 1 (traditional bucket) hydrological model.
+    """eWaterCycle implementation of Marrmot Collie River 1 (traditional bucket) hydrological model. It sets MarrmotM01 parameter with an initial value that is the mean value of the range specfied in `model parameter range file <https://github.com/wknoben/MARRMoT/blob/master/MARRMoT/Models/Parameter%20range%20files/m_01_collie1_1p_1s_parameter_ranges.m>`_.
```
```suggestion
    """eWaterCycle implementation of Marrmot Collie River 1 (traditional bucket) hydrological model. 
    
    It sets MarrmotM01 parameter with an initial value that is the mean value of the range specfied in `model parameter range file <https://github.com/wknoben/MARRMoT/blob/master/MARRMoT/Models/Parameter%20range%20files/m_01_collie1_1p_1s_parameter_ranges.m>`_.
```

See https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

Peter9192 commented on ewatercycle/models/marrmot.py:
```diff
@@ -42,42 +42,70 @@ def _generate_work_dir(work_dir: PathLike = None):
+        else:
+            raise ValueError(
+                f"Unknown container technology in CFG: {CFG['container_engine']}"
+            )
```
Why not set both? You now have repeated code (line 127 also has this same if-block). I know it's unlikely, but what if CFG is changed in between instantiation and starting the container?

SarahAlidoost replied:
Right, I applied your suggestion. However, this means that the key 'singularity_dir' should exist in CFG. See the example notebook where ewatercycle.yaml is overwritten.

Peter9192 commented on ewatercycle/models/marrmot.py:
```diff
@@ -173,7 +215,7 @@ def _create_marrmot_config(self) -> PathLike:
         )
 
-        config_file = self.work_dir / 'marrmot-m01_config.mat'
+        config_file = work_dir / Path('marrmot-m01_config.mat')
```
```suggestion
        config_file = work_dir / 'marrmot-m01_config.mat'
```
if `work_dir` is a Path instance (as the type hints suggests), then this should just work I think.

SarahAlidoost replied:
First I applied your suggestion, but then I bring it back due to mypy error:  `Unsupported left operand type for / ("_PathLike[Any]")`

Peter9192 replied:
That's strange. Can you try calling this (on the command line with you ewatercycle conda env activated)?

`mypy -c 'from pathlib import Path; Path(__file__).parent.parent / "foo" / "bar"'`

(from: https://gitter.im/python/typing?at=5d25171ff867ec1687ddc61a)

SarahAlidoost replied:
Your command line does not return an error. It seems that there is a difference in using `from pathlib import Path` vs  `from os import PathLike`. You can try the lines below in a python script and run the mypy on that script. It reproduces the error that I mentioned. 

```python
from pathlib import Path
from os import PathLike

def pathlike_fun(work_dir: PathLike):
    config_file = work_dir / 'marrmot-m01_config.mat'
    return config_file

def path_fun(work_dir: Path):
    config_file = work_dir / 'marrmot-m01_config.mat'
    return config_file
```

Peter9192 replied:
Yep, then I also get the error. @sverhoeven what is the reason to use `PathLike` rather than `Path` or perhaps `PurePath`?

sverhoeven replied:
No particular reason, it was used in the rest of the code so I continued with it.

Peter9192 replied:
Haha, ok I did the same. Perhaps we can change it here @SarahAlidoost ?

SarahAlidoost replied:
I couldn't change it only in this part of the code, because mypy returned several `Incompatible types in assignment` errors in other parts. So, I skipped changing it for now.

Peter9192 commented on ewatercycle/models/marrmot.py:
```diff
@@ -42,42 +42,70 @@ def _generate_work_dir(work_dir: PathLike = None):
+              maximum_soil_moisture_storage: float = None,
+              initial_soil_moisture_storage: float = None,
               start_time: datetime = None,
               end_time: datetime = None,
```
Start time and end time are not listed under `model.parameters`. I would expect the output of that property to match the input arguments for setup.

Peter9192 replied:
It would also be helpful if they can be passed as strings instead of datetime objects. So if `model.parameters` provides  `(start_time, "20210101")`, then `model.setup(start_time="20210102")` might be more convenient than `start_time = datetime.Datetime(2020, 1, 1))`

sverhoeven replied:
We could use `self.forcing_start_time` for the default `start_time`. 

For datetime as string I would like to standardize on [ISO8601](https://en.wikipedia.org/wiki/ISO_8601) format.
Also the forcing uses UTC while dates given setup could be timezone less, which should be handled elegantly.

SarahAlidoost replied:
now, the time of model in UTC and ISO format string can be given as 'YYYY-MM-DDTHH:MM:SS+00:00'. and start/end time are added to model.parameters.

sverhoeven replied:
What happens when you do `model.setup(start_time="2021-01-15T00:00:00+05:00")` will mat file be set to 2021-01-15 or 2021-01-16? Can a test be added

SarahAlidoost replied:
The last part i.e. `+00:00` is in `HH:MM` and shows UTC offset. In your example, the offset is 5 hours and it returns `datetime.datetime(2021, 1, 15, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=18000)))`, and it is still 2021, 1, 15, 0, 0. 
I think we should consider the UTC time as a standard (like the time in forcing data) and add a function to check that `start_time` is either in this format 'YYYY-MM-DDTHH:MM:SS+00:00' or this 'YYYY-MM-DDTHH:MM:SS' , and also to inform the user that the given `start_time` should be a UTC time.

sverhoeven replied:
Agree, instead of looking at string if timezone == UTC we could use datetime.tzinfo

```
>>> from datetime import datetime, timezone, timedelta
>>> datetime(2021, 1, 15, 0, 0, tzinfo=timezone(timedelta(seconds=18000))).tzinfo == timezone.utc
False
>>> datetime.now(timezone.utc).tzinfo == timezone.utc
True
>>> datetime.now().tzinfo == timezone.utc
False
```

SarahAlidoost replied:
I changed the [datetime iso format] to a [datetime64 iso format] i.e. 'YYYY-MM-DDTHH:MM:SSZ' that is used by [numpy](https://numpy.org/doc/stable/reference/generated/numpy.datetime_as_string.html), and xarray. I also added a util function to parse and check if time is UTC. @sverhoeven please let me know if you are happy with the changes, then we merge the PR. Thanks.

Peter9192 commented on tests/models/test_marrmotm01.py:
```diff
@@ -47,12 +27,27 @@ def forcing_file(self):
+        cfg_file, cfg_dir = model.setup()
         return model, cfg_file, cfg_dir
 
+    def test_parameters(self, model, forcing_file):
```
Are start_time and end_time considered parameters as well? Since they're part of the `setup` function, I'm inclined to see them as such.

SarahAlidoost replied:
it is fixed.

Peter9192 commented on tests/models/test_marrmotm01.py:
```diff
@@ -101,9 +94,76 @@ def test_get_value_as_xarray(self, model_with_setup):
+        # https://github.com/wknoben/MARRMoT/blob/master/BMI/Config/BMI_testcase_m01_BuffaloRiver_TN_USA.mat
+        return Path(__file__).parent / 'data' / 'BMI_testcase_m01_BuffaloRiver_TN_USA.mat'
+
+    @pytest.fixture
```
Do I understand it correctly that this fixture is specific to the class it belongs to?

sverhoeven replied:
yep, I used classes to group tests. The fixtures those tests needed can also be scoped in the class

Peter9192 commented on tests/models/test_marrmotm01.py:
```diff
@@ -101,9 +94,76 @@ def test_get_value_as_xarray(self, model_with_setup):
+        )
+        return model, cfg_file, cfg_dir
+
+    def test_setup(self, model_with_setup):
```
So this `model_with_setup` refers to `self.model_with_setup`?

sverhoeven replied:
correct, model_with_setup is a fixture scoped to the class it is in

Peter9192 reviewed (changes requested):
Looks good, and seems to be working as well for what I've tried. I have a few small comments and questions.

Peter9192 commented:
And on the notebook: maybe it would be nice to execute a few time steps to see how to actually run the model.

SarahAlidoost committed:
revise multiline docstring

Co-authored-by: Peter Kalverla <peter.kalverla@gmx.com>

SarahAlidoost committed:
Update ewatercycle/models/marrmot.py

Co-authored-by: Peter Kalverla <peter.kalverla@gmx.com>

SarahAlidoost committed:
replace datetime with date_str in setup and set iso format and UTC

SarahAlidoost committed:
add start/end time to property

SarahAlidoost committed:
refactor set_singularity_image and set_docker_image

SarahAlidoost committed:
fix error message for singularity_image.exists()

SarahAlidoost committed:
rename input arguments of _create_marrmot_config, add Path to config_file due to mypy

SarahAlidoost committed:
check if the key singularity_dir exists in CFG

SarahAlidoost committed:
change time format in tests

SarahAlidoost committed:
re-run the example noetbook

SarahAlidoost commented:
> And on the notebook: maybe it would be nice to execute a few time steps to see how to actually run the model.

I added a while loop, please note that [case -1 notebook](https://github.com/eWaterCycle/technicalPaperExampleNotebooks/blob/main/Case-1_Marrmot_Merrimack_ESMValTool.ipynb) shows how to run the model, too.

SarahAlidoost commented:
> Looks good, and seems to be working as well for what I've tried. I have a few small comments and questions.

Thank you, I addressed your comments, please have another look.

Peter9192 reviewed (approved):
Thanks @SarahAlidoost I'm happy with it as is, although the mypy error on PathLike things still annoys me.

SarahAlidoost committed:
refactor iso format and add a util function

SarahAlidoost committed:
revise a docstring

sverhoeven committed:
Add imports of marrmot to setup.py

sverhoeven committed:
Match docstring with signature

sverhoeven committed:
Typo

sverhoeven committed:
Added tests for get_time

sverhoeven committed:
Break circular import so test_util works

sverhoeven reviewed (approved):
Looks good to me. Was able to run example notebook.

My editor raised some warnings which I fixed .
To test get_time I added some tests.
